src/translation-api.js

```diff
@@ -0,0 +1,80 @@
+import { and, eq } from 'drizzle-orm'
+import { kCreateWithDocId, kSelect } from './datatype/index.js'
+import { hashObject } from './utils.js'
+
+export default class TranslationApi {
+  /** @type {Map<String,Set<import('@mapeo/schema/dist/types.js').SchemaName>>} */
+  #translatedLanguageCodeToSchemaNames = new Map()
+
+  /**
+   * @param {Object} opts
+   * @param {import('./datatype/index.js').DataType<
+   *   import('./datastore/index.js').DataStore<'config'>,
+   *   typeof import('./schema/project.js').translationTable,
+   *   'translation',
+   *   import('@mapeo/schema').Translation,
+   *   import('@mapeo/schema').TranslationValue
+   * >}  opts.dataType
+   * @param {typeof import('./schema/project.js').translationTable} opts.table
+   */
+  constructor({ dataType, table }) {
+    this.dataType = dataType
+    this.table = table
+  }
+
+  /**
+   * @param {import('@mapeo/schema').TranslationValue} value
+   */
+  async put(value) {
+    // const { message, ...identifiers } = value
+    const docId = hashObject(value)
+    let existing
+    try {
+      existing = await this.dataType.getByDocId(docId)
+    } catch (e) {
+      existing = await this.dataType[kCreateWithDocId](docId, value)
+    }
+    if (existing) {
+      await this.dataType.update(existing.versionId, value)
+    }
+    return existing.docId
+  }
+
+  /**
+   * @param {Object} value
+   * @param {String} value.docIdRef
+   * @param {String} value.fieldRef
+   * @param {String} value.schemaNameRef
+   * @param {String} value.languageCode
+   * @param {String} value.regionCode
+   */
+  async get(value) {
+    return (await this.dataType[kSelect]())
+      .where(
+        and(
+          eq(this.table.docIdRef, value.docIdRef),
+          eq(this.table.fieldRef, value.fieldRef),
+          eq(this.table.schemaNameRef, value.schemaNameRef),
+          eq(this.table.languageCode, value.languageCode),
+          // TODO: should we fallback to the first matching languageCode if regionCode doesn't match?
+          eq(this.table.regionCode, value.regionCode)
```

Comment:
yeah,  the code in [`dataType.#translate()`](https://github.com/digidem/mapeo-core-next/pull/473/files#diff-8ec3fd148a344d7225e902402078968d7097813f81c60db2fb161681b81c923dR219-R246) expects to get back all matching region codes. It also expects all fields to be returned - I think it's more efficient to return all fields and region codes in this `get()`, then parse them in memory.

However for this API standalone, maybe we can make `fieldRef` and `regionCode` optional, e.g.

```js
const filters = [
  eq(this.table.docIdRef, value.docIdRef),
  eq(this.table.schemaNameRef, value.schemaNameRef),
  eq(this.table.languageCode, value.languageCode)
]
if (value.fieldRef) {
  filters.push(eq(this.table.fieldRef, value.fieldRef))
}
if (value.regionCode) {
  filters.push(eq(this.table.regionCode, value.regionCode))
}
return (await this.dataType[kSelect]())
       .where(and.apply(null, filters))
```

Reply:
how can `fieldRef` be optional? if that's the case then when requesting a translation for the key 'monitor-name' in a `Field`, you'll get every translated key instead of just what you want

Reply:
In the actual usage [dataType.#translate()](https://github.com/digidem/mapeo-core-next/pull/473/files#diff-8ec3fd148a344d7225e902402078968d7097813f81c60db2fb161681b81c923dR219-R246) you can see that in order to translate a record, we need to check whether each field has a translation. I thought it was more efficient to make a single query for all fields for a particular docId, since we will need all of them, rather than make a query for each field.

Reply:
It's because we need to check every field for every record for a translation, for every read of the database. The two strategies in the implementation to limit the performance impact of this are:

1. Index which datatypes have any translations to a particular languageCode
2. If, from the index, we do know that e.g. `preset` datatype has at least one translation to `es` languageCode, then we do one query for each document, by `docIdRef`, `schemaIdRef` and `languageCode`, and then translate any fields from the returned rows.

Reply:
Yeah, that makes a lot of sense